### PR TITLE
[misc] feat: Add bidirectional MLM <-> MBridge config translation script

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary

- **Add `scripts/translate_mlm_to_bridge.py`** — a bidirectional translation script that converts between Megatron-LM (`pretrain_gpt.py`) CLI arguments and Megatron Bridge (`run_recipe.py`) Hydra-style overrides.
- Supports **MLM→Bridge** (default) and **Bridge→MLM** (`--reverse`) directions.
- Outputs as CLI overrides, standalone recipe file, or full `torchrun` command in both directions.

## Motivation

Running correlation tests between Megatron-LM and Megatron Bridge requires matching configurations precisely across two very different config systems (argparse flags vs Hydra overrides on dataclass recipes). This script automates the translation, eliminating manual errors and making it easy to verify that both frameworks receive identical model/training parameters.

## MLM → Bridge Translation

Translates Megatron-LM `pretrain_gpt.py` arguments (from YAML or CLI) into Bridge recipe overrides:

```bash
# From a YAML config file
python scripts/translate_mlm_to_bridge.py config.yaml

# From inline args
python scripts/translate_mlm_to_bridge.py -- --num-layers 32 --hidden-size 4096 --bf16

# Emit a full torchrun command
python scripts/translate_mlm_to_bridge.py config.yaml --format command --nproc 8
```

## Bridge → MLM Translation

Translates Bridge Hydra overrides into Megatron-LM CLI arguments:

```bash
# From CLI override string
python scripts/translate_mlm_to_bridge.py --reverse \
  "model.num_layers=32 model.hidden_size=4096 mixed_precision=bf16_mixed"

# From a Bridge YAML/override file
python scripts/translate_mlm_to_bridge.py --reverse bridge_config.yaml

# Emit a full torchrun command for pretrain_gpt.py
python scripts/translate_mlm_to_bridge.py --reverse \
  --format command --nproc 8 \
  "model.num_layers=32 model.hidden_size=4096"
```

## Key mappings handled

| Category | Examples |
|---|---|
| Model architecture | `num_layers`, `hidden_size`, `num_attention_heads`, `ffn_hidden_size`, `normalization`, `position_embedding_type`, `rotary_base` |
| Activation | `swiglu` ↔ `activation_func=silu` + `gated_linear_unit=true`, `squared-relu` ↔ `activation_func=squared_relu` |
| Training | `micro_batch_size`, `global_batch_size`, `train_iters`, `lr`, `weight_decay`, `clip_grad`, `seed` |
| Parallelism | `tensor_model_parallel_size`, `pipeline_model_parallel_size`, `context_parallel_size`, `sequence_parallel`, `use_distributed_optimizer` |
| Precision | `bf16` ↔ `bf16_mixed`, `fp16` ↔ `fp16_mixed` |
| Tokenizer | `tokenizer_type`, `tokenizer_model`, `vocab_size` |

## Test plan

- [ ] Existing unit tests pass
- [ ] CI triggered with `/ok to test`
- [x] Verified round-trip translation preserves key parameters
- [x] Verified loss correlation between MLM and Bridge with matched parameters using this script

### CODE REVIEW

File: scripts/translate_mlm_to_bridge.py [hunk 1]
```diff
@@ -0,0 +1,1635 @@
...
+      --args "train.train_iters=1000 model.tensor_model_parallel_size=2"
+
+  # Bridge → MLM: recipe only (all defaults)
+  python scripts/translate_mlm_to_bridge.py --reverse \\
+      --recipe vanilla_gpt_pretrain_config
+
+  # Bridge → MLM: overrides only (no recipe)
+  python scripts/translate_mlm_to_bridge.py --reverse \\
+      --args "model.num_layers=32 model.activation_func=silu model.gated_linear_unit=true"
+"""
```
Comment (cuichenx) [inline scripts/translate_mlm_to_bridge.py:52]: clarify --emit output is stdout unless -o is specified

File: docs/megatron-lm-to-megatron-bridge.md [hunk 2]
```diff
@@ -2,6 +2,85 @@
...
+python scripts/translate_mlm_to_bridge.py --yaml model_configs/DeepSeek-V3.yaml
+
+# From inline CLI args
+python scripts/translate_mlm_to_bridge.py \
+    --args "--num-layers 32 --hidden-size 4096 --num-attention-heads 32 --bf16 --swiglu"
+
+# Emit a standalone Bridge recipe Python file
+python scripts/translate_mlm_to_bridge.py \
+    --yaml DeepSeek-V3.yaml --emit recipe --recipe-name deepseek_v3
+```
```
Comment (cuichenx) [inline docs/megatron-lm-to-megatron-bridge.md:26]: clarify default output is stdout and a output file can be specified

File: docs/megatron-lm-to-megatron-bridge.md [hunk 3]
```diff
@@ -2,6 +2,85 @@
...
+| `--save-interval N` | `checkpoint.save_interval=N` | |
+| `--tokenizer-type T` | `tokenizer.tokenizer_type=T` | |
+| `--tokenizer-model M` | `tokenizer.tokenizer_model=M` | |
+| `--normalization N` | `model.normalization=N` | |
+| `--position-embedding-type T` | `model.position_embedding_type=T` | |
+| `--rotary-base N` | `model.rotary_base=N` | |
+| `--num-experts N` | `model.num_moe_experts=N` | |
+| `--moe-router-topk K` | `model.moe_router_topk=K` | |
+
+Flags not present in Bridge (e.g., `--use-mcore-models`, `--use-flash-attn`, `--mock-data`) are silently skipped with a comment. Unknown flags are listed in a separate section so you can handle them manually.
```
Comment (cuichenx) [inline docs/megatron-lm-to-megatron-bridge.md]: --mock-data should be translated to `dataset.mock=True` for pretraining right?

File: docs/megatron-lm-to-megatron-bridge.md [hunk 4]
```diff
@@ -2,6 +2,85 @@
...
+# From a YAML config file (MODEL_ARGS section)
+python scripts/translate_mlm_to_bridge.py --yaml model_configs/DeepSeek-V3.yaml
+
+# From inline CLI args
+python scripts/translate_mlm_to_bridge.py \
+    --args "--num-layers 32 --hidden-size 4096 --num-attention-heads 32 --bf16 --swiglu"
+
+# Emit a standalone Bridge recipe Python file
+python scripts/translate_mlm_to_bridge.py \
+    --yaml DeepSeek-V3.yaml --emit recipe --recipe-name deepseek_v3
```
Comment (yaoyu-33) [inline docs/megatron-lm-to-megatron-bridge.md:21]: clarify what emit reicpe do? also this seems to be wrong "emit_recipe"?

File: src/megatron/bridge/training/utils/omegaconf_utils.py [hunk 5]
```diff
@@ -36,6 +38,62 @@
...
+    parts = name.rsplit(".", 1)
+    if len(parts) == 2:
+        try:
+            module = importlib.import_module(parts[0])
+            return getattr(module, parts[1])
+        except (ImportError, AttributeError):
+            pass
+    raise ValueError(
+        f"Unknown activation function: '{name}'. Known names: {sorted(n for n in ACTIVATION_FUNC_MAP if '.' not in n)}"
+    )
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply src/megatron/bridge/training/utils/omegaconf_utils.py]: Fixed in 70aa3e6e.

File: scripts/translate_mlm_to_bridge.py (hunk 1 above)
Comment (yaoyu-33) [reply scripts/translate_mlm_to_bridge.py:52]: Fixed in 70aa3e6e.

File: docs/megatron-lm-to-megatron-bridge.md (hunk 2 above)
Comment (yaoyu-33) [reply docs/megatron-lm-to-megatron-bridge.md:26]: Fixed in 70aa3e6e.

File: docs/megatron-lm-to-megatron-bridge.md (hunk 3 above)
Comment (yaoyu-33) [reply docs/megatron-lm-to-megatron-bridge.md]: Fixed in 70aa3e6e.

File: docs/megatron-lm-to-megatron-bridge.md (hunk 4 above)
Comment (yaoyu-33) [reply docs/megatron-lm-to-megatron-bridge.md:21]: Fixed in 70aa3e6e.

File: scripts/translate_mlm_to_bridge.py [hunk 6]
```diff
@@ -0,0 +1,1635 @@
...
+        if isinstance(parsed, (list, tuple, dict)):
+            return parsed
+        return parsed
+    except (ValueError, SyntaxError):
+        pass
+
+    return _try_numeric(val)
+
+
+def parse_yaml_config(path: str) -> dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply scripts/translate_mlm_to_bridge.py]: Fixed in 70aa3e6e.

File: src/megatron/bridge/training/utils/omegaconf_utils.py [hunk 7]
```diff
@@ -486,6 +503,14 @@ def _apply_overrides(config_obj: DataclassInstance, overrides_dict: Dict[str, An
                         logger.warning(f"Could not convert string '{value}' back to torch.dtype")
                         final_value = value
 
+                # Restore serialized callable fields (e.g. "relu" → F.relu)
+                if key in _SERIALIZABLE_CALLABLE_FIELDS and isinstance(final_value, str):
+                    restored = str_to_callable(final_value)
+                    if restored is not None:
+                        final_value = restored
+                    else:
+                        logger.warning(f"Could not restore callable for {key}='{final_value}'; keeping string")
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply src/megatron/bridge/training/utils/omegaconf_utils.py:513]: Fixed in 70aa3e6e.

File: src/megatron/bridge/utils/activation_map.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import torch
+import torch.nn.functional as F
+from megatron.core.activations import fast_gelu, squared_relu
+
+
+logger = logging.getLogger(__name__)
+
+# Canonical name -> callable.
+# Short names come first; fully-qualified aliases are added below.
+ACTIVATION_FUNC_MAP: dict[str, Callable] = {
```
Comment (maanug-nv) [inline src/megatron/bridge/utils/activation_map.py:35]: we should document this so that users know they can use these string names via hydra or yaml
